dvc/remote/local.py

```diff
@@ -348,48 +428,121 @@ def _process(
+        dir_plans = self._get_plans(download, remote, dir_status, status)
+        file_plans = self._get_plans(download, remote, file_status, status)
 
-        if len(plans[0]) == 0:
+        if len(dir_plans[0]) + len(file_plans[0]) == 0:
             return 0
 
-        if jobs > 1:
-            with ThreadPoolExecutor(max_workers=jobs) as executor:
-                fails = sum(executor.map(func, *plans))
-        else:
-            fails = sum(map(func, *plans))
+        with ThreadPoolExecutor(max_workers=jobs) as executor:
+            if download:
+                fails = sum(executor.map(func, *dir_plans))
+                fails += sum(executor.map(func, *file_plans))
+            else:
+                # for uploads, push files first, and any .dir files last
+
+                file_futures = {}
+                for from_info, to_info, name in zip(*file_plans):
+                    file_futures[to_info] = executor.submit(
+                        func, from_info, to_info, name
+                    )
+                dir_futures = {}
+                for from_info, to_info, name in zip(*dir_plans):
+                    wait_futures = {
+                        future
+                        for file_path, future in file_futures.items()
+                        if file_path in dir_paths[to_info]
+                    }
+                    dir_futures[to_info] = executor.submit(
+                        self._dir_upload,
+                        func,
+                        wait_futures,
+                        from_info,
+                        to_info,
+                        name,
+                    )
+                fails = sum(
+                    future.result()
+                    for future in concat(
+                        file_futures.values(), dir_futures.values()
+                    )
+                )
 
         if fails:
             if download:
+                remote.index.invalidate()
                 raise DownloadError(fails)
             raise UploadError(fails)
+        elif not download:
+            pushed_checksums = list(
+                map(self.path_to_checksum, dir_plans[0] + file_plans[0])
+            )
+            logger.debug(
+                "Adding {} pushed checksums to index".format(
+                    len(pushed_checksums)
+                )
+            )
+            remote.index.update(pushed_checksums)
+            remote.index.save()
+
+        return len(dir_plans[0]) + len(file_plans[0])
+
+    def _dir_upload(self, func, futures, from_info, to_info, name):
+        for future in as_completed(futures):
+            if future.result():
+                # do not upload this .dir file if any file in this
+                # directory failed to upload
+                logger.debug(
+                    "failed to upload full contents of '{}', "
+                    "aborting .dir file upload".format(name)
+                )
+                logger.error(
```

Comment:
include the previous logger here?

Reply:
The error will already have been caught and logged in the exception handler for the failed upload thread. I'm not sure if it's worth storing the exception info so it can be referenced across threads and logged here?

Reply:
I think my question was mostly about the duplication logger.debug + logger.error - do we really need both?

Reply:
Ah, my reasoning for separating it was that ".dir file was not uploaded" is relevant to the user, but the specific reason why in this case seemed less relevant outside of of a verbose/debugging output context

Reply:
kk, up to you, I would personally try something like:

`upload {} to {}, one or more files failed` ?

it does not hurt to show some reason, will help us debug faster, simplifies code (best code == no code :) )